code/game/turfs/turf.dm

```diff
@@ -538,8 +563,44 @@
 			if(nutri_check.nutriment_factor >0)
 				M.reagents.remove_reagent(R.type, min(R.volume, 10))
 
-//Whatever happens after high temperature fire dies out or thermite reaction works.
-//Should return new turf
+/turf/proc/set_heat_damage_amount(temperature)
+	var/obj/structure/window/W = locate() in range(1, src)
+	if(W)
+		W.adjacent_fire_act(temperature)
+	var/obj/machinery/door/D = locate() in range(1, src)
+	if(D)
+		D.adjacent_fire_act(temperature)
+	if(to_be_destroyed && !changing_turf)
+		damage_amount = clamp((temperature - heat_capacity)/10000, 1, 15)
+		turf_take_heat_damage(damage_amount)
+		heat_capacity = max(heat_capacity - (damage_amount * 10), 0)
+	else
+		to_be_destroyed = FALSE
+		max_fire_temperature_sustained = 0
+
+/turf/proc/turf_take_heat_damage(damage_amount)
+	if(QDELETED(src))
+		stack_trace("[src] taking damage after deletion")
+		return
+	if(turf_heat_resistance <= 0)
+		return
+	turf_heat_resistance = max(turf_heat_resistance - damage_amount, 0)
+	update_overlays()
+	if(turf_heat_resistance <= 0)
+		Melt()
```

Comment:
This is a bit strange.
In general health is divided into a variable holding the max health value and another holding the current health one.
This one seems to have a single variable. Does temperature do irreversible damage to turfs?